Raise Error for Unsupported Bundler Version

kbukum1 opened the pull request:
### What are you trying to accomplish?
This PR introduces logic to raise an error when the version of Bundler is unsupported. The change is controlled by the `bundler_v1_unsupported_error` feature flag, allowing the error to be activated only when the flag is enabled.

### Anything you want to highlight for special attention from reviewers?
The feature flag `bundler_v1_unsupported_error` controls whether the error is raised for unsupported Bundler v1 versions. 

### How will you know you've accomplished your goal?
The goal is achieved if:
- The error is raised correctly when the feature flag is enabled and Bundler v1 is unsupported.
- The error will be displayed on the Dependabot Insights page and the Security Alerts page.
- The behavior is properly controlled by the feature flag.

### Checklist
- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

kbukum1 commented on common/lib/dependabot/notices.rb:
```diff
@@ -144,40 +128,6 @@ def self.generate_pm_deprecation_notice(package_manager)
-        show_alert: true
-      )
-    end
-
```
This was originally created to follow the notice structure when implementing the deprecation feature. However, after implementing the unsupported error, I realized we don’t need a similar structure to the deprecation notice. We already have a specialized error (`Dependabot::ToolVersionNotSupported`) used in other cases, so we are generalizing that for unsupported versions.

Reference code: https://github.com/dependabot/dependabot-core/blob/main/common/lib/dependabot/errors.rb#L435

kbukum1 commented on common/lib/dependabot/notices.rb:
```diff
@@ -94,22 +94,6 @@ def self.generate_supported_versions_description(supported_versions, support_lat
-
-      generate_pm_unsupported_notice(package_manager)
-    end
-
```
This was originally created to follow the notice structure when implementing the deprecation feature. However, after implementing the unsupported error, I realized we don’t need a similar structure to the deprecation notice. We already have a specialized error (Dependabot::ToolVersionNotSupported) used in other cases, so we are generalizing that for unsupported versions.

Reference code: https://github.com/dependabot/dependabot-core/blob/main/common/lib/dependabot/errors.rb#L435

kbukum1 commented on common/lib/dependabot/package_manager.rb:
```diff
@@ -58,18 +58,36 @@ def deprecated?
+
+      raise ToolVersionNotSupported.new(name, version.to_s, supported_versions_message)
+    end
+
```
We are dynamically creating supported version string here that is going to be shown in the error.

kbukum1 commented on common/lib/dependabot/package_manager.rb:
```diff
@@ -58,18 +58,36 @@ def deprecated?
       return true if unsupported_versions.include?(version)
 
       supported_versions = self.supported_versions
-      return version < supported_versions.first if supported_versions.any?
```
This method is written when doing deprecation warning however it is not used since we didn't implemented yet at that time unsupported error. We are refining this method to fit the case

kbukum1 commented on common/lib/dependabot/package_manager.rb:
```diff
@@ -58,18 +58,36 @@ def deprecated?
     # @example
     #   package_manager.unsupported? #=> false
     sig { returns(T::Boolean) }
     def unsupported?
```
Just note that this method written in general however the implementation can override this so bundler v1 package manager currently implementing this.

kbukum1 commented on bundler/lib/dependabot/bundler/package_manager.rb:
```diff
@@ -44,9 +43,14 @@ def initialize(version)
+        return false unless Dependabot::Experiments.enabled?(:bundler_v1_unsupported_error)
+
+        # Determine if the Bundler version is unsupported.
+        version < supported_versions.first
```
We expect the version to be lower then the first element of supported version if it is not supported. In other cases we are unable to determine if it is not supported.

kbukum1 commented on bundler/lib/dependabot/bundler/package_manager.rb:
```diff
@@ -44,9 +43,14 @@ def initialize(version)
       def unsupported?
-        !deprecated? && version < supported_versions.first
+        # Check if the feature flag for Bundler v1 unsupported error is enabled.
+        return false unless Dependabot::Experiments.enabled?(:bundler_v1_unsupported_error)
```
This feature flag is going to be used to enable not supported version for bundler v1 when the date arrives.

kbukum1 commented on common/spec/dependabot/notices_spec.rb:
```diff
@@ -96,86 +114,6 @@ def initialize(name:, version:, deprecated_versions: [], unsupported_versions: [
     end
   end
 
-  describe ".generate_support_notice" do
```
Not used generate support notice method are removed after finding out that we don't need to create notice. Instead we are using general error flow to raise unsupported error.

kbukum1 commented on common/spec/dependabot/notices_spec.rb:
```diff
@@ -204,33 +142,4 @@ def initialize(name:, version:, deprecated_versions: [], unsupported_versions: [
     end
   end
-
-  describe ".generate_pm_unsupported_notice" do
```
Not used generate pm unsupported notice method is removed after finding out that we don't need to create notice. Instead we are using general error flow to raise unsupported error.

kbukum1 commented on bundler/lib/dependabot/bundler/package_manager.rb:
```diff
@@ -9,9 +9,8 @@ module Dependabot
-      Version.new("2")
-    ].freeze, T::Array[Dependabot::Version])
+    # Keep versions in ascending order
+    SUPPORTED_BUNDLER_VERSIONS = T.let([Version.new("2")].freeze, T::Array[Dependabot::Version])
```
Keeping the array in ascending order is important because we compare the first element of `supported_versions` with the current version (`version < supported_versions.first`) to determine if it is unsupported.

kbukum1 replied:
We added specs to check if this array is ordered, so we don't need to run sorting.

landongrindheim replied:
_Assuming_ that we'd only support versions greater than X, we could remove the requirement of having developers manually sorting by checking that a version is greater than all versions in this collection.

That would fall apart if we supported, say PackageManager `v1` and `v3`, but not `v2`. However, that would also fall apart here.

kbukum1 replied:
applied the change to compare with each of them.

kbukum1 commented on common/spec/dependabot/notices_spec.rb:
```diff
@@ -23,6 +24,23 @@ def initialize(name:, version:, deprecated_versions: [], unsupported_versions: [
+  sig { override.returns(T::Boolean) }
+  def unsupported?
+    # Check if the feature flag for Bundler v1 unsupported error is enabled.
+    return false unless Dependabot::Experiments.enabled?(:bundler_v1_unsupported_error)
```
The feature flag going to be activated when we are not supporting the version anymore.

kbukum1 commented:
Tested on operations locally

When feature flag is on

<img width="1421" alt="Screenshot 2024-09-16 at 12 56 42 PM" src="https://github.com/user-attachments/assets/252e2900-2811-4683-81ff-0eba12b280bc">

When feature flag is off
<img width="735" alt="Screenshot 2024-09-16 at 1 03 04 PM" src="https://github.com/user-attachments/assets/cf27036b-2ca6-4884-baa2-b50480ec3451">

kbukum1 commented:
Just note that need to test feature flag on a test repository after the deployment.

kbukum1 commented on updater/spec/dependabot/updater/operations/create_security_update_pull_request_spec.rb:
```diff
@@ -15,23 +15,6 @@
 require "dependabot/bundler"
 
-# Stub PackageManagerBase
-class StubPackageManager < Dependabot::PackageManagerBase
```
Moved into `DummyPkgHelpers`

kbukum1 commented on updater/spec/dependabot/updater/operations/refresh_version_update_pull_request_spec.rb:
```diff
@@ -16,23 +16,6 @@
 require "dependabot/bundler"
 
-# Stub PackageManagerBase
-class StubPackageManager < Dependabot::PackageManagerBase
```
Moved into `DummyPkgHelpers`

kbukum1 commented on updater/spec/dependabot/updater/operations/update_all_versions_spec.rb:
```diff
@@ -18,23 +18,6 @@
 require "dependabot/bundler"
 
-# Stub PackageManagerBase
-class StubPackageManager < Dependabot::PackageManagerBase
```
Moved into `DummyPkgHelpers`

landongrindheim commented on bundler/lib/dependabot/bundler/package_manager.rb:
```diff
@@ -40,13 +39,13 @@ def initialize(version)
+        # Check if the feature flag for Bundler v1 unsupported error is enabled.
+        return false unless Dependabot::Experiments.enabled?(:bundler_v1_unsupported_error)
+
+        # Determine if the Bundler version is unsupported.
+        supported_versions.any? && version < supported_versions.first
```
💅 What do you think of something like 👇?

```suggestion
        supported_versions.all? { |supported| supported > version }
```

landongrindheim commented on bundler/spec/dependabot/bundler/package_manager_spec.rb:
```diff
@@ -50,66 +50,96 @@
+      let(:feature_flag_enabled) { true }
 
-      it "returns false" do
+      it "returns false, as unsupported takes precedence" do
```
🤔 I can see room for confusion if `#deprecated?` goes from `true` to `false` when support is dropped, but I think this is correct.

landongrindheim commented on common/lib/dependabot/package_manager.rb:
```diff
@@ -64,12 +67,23 @@ def deprecated?
     sig { returns(T::Boolean) }
     def unsupported?
-      return true if unsupported_versions.include?(version)
+      false
```
⚠️ I'm concerned that this might have unintended consequences. Have you audited all package managers to ensure they'll still be supported when this merges?

Outside of the scope of this work, but `#supported?`, with a default of `true` would be a little easier _for me_ to reason about 😅

kbukum1 replied:
Bundler was the first package manager to implement this feature, and I initially added the `unsupported?` method, though it wasn’t used at the time since it was part of a future task. For other package managers, we assume the version is supported by default, unless the `unsupported?` condition is explicitly implemented in their respective versions.

landongrindheim commented on updater/lib/dependabot/updater/group_update_creation.rb:
```diff
@@ -241,6 +241,9 @@ def compile_updates_for(dependency, dependency_files, group) # rubocop:disable M
           return []
         end
 
+        # Raise an error if the package manager version is unsupported
+        dependency_snapshot.package_manager&.raise_if_unsupported!
```
🔎 Is the `&.` needed here for changes which happen across ecosystems?

kbukum1 replied:
Yes, because currently other eco-system don't have package_manager implementation. We always need package manager object from eco-system to be set to do this operations.

landongrindheim commented on updater/spec/support/dummy_pkg_helpers.rb:
```diff
@@ -61,4 +62,38 @@ def create_temporary_content_directory(fixture:, directory: "/", state: "origina
+
+    sig { override.returns(T::Boolean) }
+    def unsupported?
+      # Check if the feature flag for Bundler v1 unsupported error is enabled.
+      return false unless Dependabot::Experiments.enabled?(:bundler_v1_unsupported_error)
+
+      # Determine if the Bundler version is unsupported.
+      version < supported_versions.first
```
🔎 I'm okay with this landing _now_, but I think it's preferable to exclude Bundler-specific concerns in this generic stub.

```suggestion
      # Check if the feature flag for Bundler v1 unsupported error is enabled.
      return false unless name == "bundler" && Dependabot::Experiments.enabled?(:bundler_v1_unsupported_error)

      version < supported_versions.first
```

kbukum1 commented:
@landongrindheim 

Thank you for the review! I’ve addressed all the comments and provided responses. 

Just a quick note: Currently, only Bundler has implemented this package manager (also only deprecation is active), but we plan to extend it to all ecosystems in the future.
